src/fn_call.rs

```diff
@@ -71,8 +76,195 @@ pub trait EvalContextExt<'a, 'mir, 'tcx: 'a + 'mir>: crate::MiriEvalContextExt<'
+
+                let box_me_up_ptr = alloc.read_ptr_sized(
+                    this,
+                    vtable_ptr.offset(this.pointer_size() * 3, this)?
+                )?.to_ptr()?;
+
+                // Get the actual function instance
+                let box_me_up_fn = this.memory().get_fn(box_me_up_ptr)?;
+                let box_me_up_mir = this.load_mir(box_me_up_fn.def)?;
+
+                // Extract the signature
+                // We know that there are no HRBTs here, so it's fine to use
+                // skip_binder
+                let fn_sig_temp = box_me_up_fn.ty(*this.tcx).fn_sig(*this.tcx);
+                let fn_sig = fn_sig_temp.skip_binder();
+
+                // This is the layout of '*mut (dyn Any + Send)', which
+                // is the return type of 'box_me_up'
+                let dyn_ptr_layout = this.layout_of(fn_sig.output())?;
+
+                // We allocate space to store the return value of box_me_up:
+                // '*mut (dyn Any + Send)', which is a fat 
+                
+                let temp_ptr = this.allocate(dyn_ptr_layout, MiriMemoryKind::UnwindHelper.into());
+
+                // Keep track of our current frame
+                // This allows us to step throgh the exection of 'box_me_up',
+                // exiting when we get back to this frame
+                let cur_frame = this.cur_frame();
+
+                this.push_stack_frame(
+                    box_me_up_fn,
+                    box_me_up_mir.span,
+                    box_me_up_mir,
+                    Some(temp_ptr.into()),
+                    StackPopCleanup::None { cleanup: true }
+                )?;
+
+                let mut args = this.frame().mir.args_iter();
+                let arg_0 = this.eval_place(&mir::Place::Base(mir::PlaceBase::Local(args.next().unwrap())))?;
+                this.write_scalar(data_ptr, arg_0)?;
+
+                // Step through execution of 'box_me_up'
+                // We know that we're finished when our stack depth
+                // returns to where it was before.
+                //
+                // Note that everything will get completely screwed up
+                // if 'box_me_up' panics. This is fine, since this
+                // function should never panic, as it's part of the core
+                // panic handling infrastructure
+                //
+                // Normally, we would just let Miri drive
+                // the execution of this stack frame.
+                // However, we need to access its return value
+                // in order to properly unwind.
+                //
+                // When we 'return' from '__rustc_start_panic',
+                // we need to be executing the panic catch handler.
+                // Therefore, we take care all all of the unwinding logic
+                // here, instead of letting the Miri main loop do it
+                while this.cur_frame() != cur_frame {
+                    this.step()?;
```

Comment:
assert here that we are always getting a `MoreToDo`

Reply:
This sadly makes it impossible to inspect execution during unwinding in [priroda](https://github.com/oli-obk/priroda). Maybe add a StackPopCleanup variant taking an arbitrary closure to evaluate on stack pop and put the code below in that closure.

Reply:
This "just" steps through `box_me_up`, right, not through the actual unwind? But yeah, would still be nice to avoid nested interpreter loops. We have another one of these in the TLS cleanup and I don't like it.

Reply:
Unfortunately, I didn't see a way of avoiding the nested loop while still getting access to the return value of `box_me_up`

Reply:
could you put it into a field on the `Frame` or as suggested by @bjorn3 create a new `StackPopCleanup` that you'd use for the newly pushed frame?

Reply:
The work that has to happen between `box_me_up` returning and jumping to the caller's unwind continuation is... what?

Using closure for this is often hard as they tend to capture lots of stuff. And in particular when doing this in a loop, like we need for TLS d'tors, you get a headache quickly. Maybe having a `Machine` associated type for "other action on stack pop", and an associated hook in the trait as well, would work better because one would have to be a bit more explicit about what the data needed for those hooks is?

Reply:
I no longer explicitly step through `box_me_up`